tracer.go

```diff
@@ -0,0 +1,223 @@
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.maxAttempts = maxAttempts
+}
+
+func (t *traceWriter) SetSleepInterval(sleepInterval time.Duration) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.sleepInterval = sleepInterval
+}
+
+func (t *traceWriter) Trace(traceId []byte) {
+	var (
+		timestamp time.Time
+		activity  string
+		source    string
+		elapsed   int
+		thread    string
+	)
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	fetchAttempts := 1
+	if t.maxAttempts > 0 {
+		fetchAttempts = t.maxAttempts
+	}
+
+	isDone := false
+	for i := 0; i < fetchAttempts; i++ {
+		var duration int
+
+		iter := t.session.control.query(`SELECT duration
+			FROM system_traces.sessions
+			WHERE session_id = ?`, traceId)
+		iter.Scan(&duration)
+		if duration > 0 {
+			isDone = true
+		}
+
+		if err := iter.Close(); err != nil {
+			fmt.Fprintln(t.w, "Error:", err)
+			return
+		}
+
+		if isDone || i == fetchAttempts-1 {
+			break
+		}
+
+		time.Sleep(t.sleepInterval)
+	}
+	if !isDone {
+		fmt.Fprintln(t.w, "Error: failed to wait tracing to complete. !!! Tracing is incomplete !!!")
+	}
+
+	var (
+		coordinator string
+		duration    int
+	)
+
+	iter := t.session.control.query(`SELECT coordinator, duration
+		FROM system_traces.sessions
+		WHERE session_id = ?`, traceId)
+
+	iter.Scan(&coordinator, &duration)
+	if err := iter.Close(); err != nil {
+		fmt.Fprintln(t.w, "Error:", err)
+		return
+	}
+
+	fmt.Fprintf(t.w, "Tracing session %016x (coordinator: %s, duration: %v):\n",
+		traceId, coordinator, time.Duration(duration)*time.Microsecond)
+
+	iter = t.session.control.query(`SELECT event_id, activity, source, source_elapsed, thread
+			FROM system_traces.events
+			WHERE session_id = ?`, traceId)
+
+	for iter.Scan(&timestamp, &activity, &source, &elapsed, &thread) {
+		fmt.Fprintf(t.w, "%s: %s [%s] (source: %s, elapsed: %d)\n",
+			timestamp.Format("2006/01/02 15:04:05.999999"), activity, thread, source, elapsed)
+	}
+
+	if err := iter.Close(); err != nil {
+		fmt.Fprintln(t.w, "Error:", err)
+	}
+}
+
+type tracer struct {
+	session  *Session
+	traceIDs [][]byte
+	mu       sync.Mutex
+}
+
+func NewTracer(session *Session) *tracer {
+	return &tracer{session: session}
+}
+
+func (t *tracer) Trace(traceId []byte) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.traceIDs = append(t.traceIDs, traceId)
+}
+
+func (t *tracer) AllTraceIDs() [][]byte {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	return t.traceIDs
+}
+
+func (t *tracer) IsReady(traceId []byte) (bool, error) {
+	iter := t.session.control.query(`SELECT event_id, activity, source, source_elapsed, thread
+		FROM system_traces.events
+		WHERE session_id = ?`, traceId)
+
+	var (
+		timestamp time.Time
+		activity  string
+		source    string
+		elapsed   int
+		thread    string
+	)
+
+	for iter.Scan(&timestamp, &activity, &source, &elapsed, &thread) {
+		if strings.Contains(activity, "preparing a result") {
+			return true, nil
+		}
+	}
+
+	if err := iter.Close(); err != nil {
+		return false, err
+	}
+
+	return false, nil
```

Comment:
Again, I don't see you using `duration` here from `system_traces.sessions`

Reply:
I changed it in one place instead of two, sorry